### DISCUSSION TITLE
OSPI PSRAM support for B-U585I-IOT02A

### DISCUSSION THREAD

**Discussion by mariopaja** [zephyrproject-rtos/zephyr#70844] [Q&A] [answered]
Is there any previous work, or has someone started to work on the support of OSPI PSRAM that is on B-U585I-IOT02A board

**Comment by FRASTM** [comment, marked as answer]
hi @mariopaja 
The support of the OSPI PSRAM is not developed. 
OctoSPI 1 bus of the b_u585i_iot01a disco kit is connected to the PSRAM but the  flash_stm32_ospi driver is not adapted for accessing the PSRAM device

**Reply by mariopaja** [reply]
@FRASTM 
A possible OSPI PSRAM would have to chip specific like the implementation of _memc_mcux_flexspi_xxxxxx_, right?

**Reply by erwango** [reply]
@mariopaja Exactly

**Comment by mariopaja** [comment]
@erwango @FRASTM 

I just saw the new MSPI API  and I was wondering if it can be used to drive the PSRAM on the B-U585I-IOT02A board?

I understand that **_memc_mspi_aps6408l.c_** would be needed, but would this (mspi_emul.c) remove the need fo calling HAL specific functions for driving the PSRAM?

Or am I understanding the funksionality of this API wrong?

**Comment by srberard** [comment]
@FRASTM and @erwango 
Similarly to @mariopaja above, I'm trying to unlock the PSRAM on the B-U585-IOT02A board.  Has there been any work towards getting this to work?  I've looked at it seems there is support for this in the ST BSP/HAL.  There is also a driver here:  https://github.com/STMicroelectronics/stm32-aps6408

I'm not sure of the effort of porting this or using this in Zephyr.  I plan to explore this a bit but wanted to check to see if anyone else has started this work.

**Reply by erwango** [reply]
@srberard This feature is now supported on N6 through `drivers/memc/memc_stm32_xspi_psram.c`.
On U5, this would take some adaptation to make it working through OSPI driver and, on some latest U5 supporting HSPI, this exact driver would have to be updated with minor tweaks.

**Reply by mariopaja** [reply]
@srberard I have done an implementation of the psram which i ported from the example in STM Cube. It was not difficult at all. We also used the same PSRAM as the devboard and I didnt need to change much. The implementation is in Zephyr v3.7 LTS

@erwango I didn't know that the PSRAM is implemented for the N6. I will have a look into it and maybe I will come with the implementation of the U5

**Reply by srberard** [reply]
@mariopaja Thanks for this. Is there anything special I need to define in Device Tree or enable to get this driver working?

I've been mostly working on 4.1, but will try 3.7 in the morning.

**Reply by mariopaja** [reply]
@srberard The zephyr version should not really matter. 

I based the driver partially on the OSPI flash and I ended up having a custom driver for the aps6408l.

In my driver a linker is needed to specify where the variables should be saved. Additionally the user must specify which variables will be stored in the psram. It is not the best implementation but for our case worked

It is more or less the same sample as the stmcube

**Reply by srberard** [reply]
Ah, sorry, I misunderstood. I thought you had the driver checked into the Zephyr code base. I understand now. I'll check out the STM Cube samples.

**Reply by dkouba-atym** [reply]
@mariopaja would you be willing to share your driver implementation with us? I'm looking to use the PSRAM for a demo I am slated to deliver by end of month, and a head start would be greatly appreciated.

**Reply by mariopaja** [reply]
@dkouba-atym Unfortunately I cannot share it. I will ask if I can open a PR for this matter but I cannot promise anything.

**Reply by mariopaja** [reply]
@dkouba-atym @srberard I have already received permission to push the driver upstream. I will make some changes to the driver and hopefully push next week a draft.

**Reply by mariopaja** [reply]
@dkouba-atym @srberard please find here the first draft https://github.com/zephyrproject-rtos/zephyr/pull/92381
